### PR TITLE
Adds a new mod module - mind swap - to illegal tech

### PR DESCRIPTION
**Pull request by Odairu**
## About The Pull Request
This module allows for the at-will body swapping between a suit AI and a player mob. At face value its not really useful for real mechanical purposes as, neither player will be able to be able to really do anything extra that their counterpart couldn't do. I just personally think its cool and could add some new cool scenarios for when wearing a suit AI
## Why It's Good For The Game
I've been told its hot and cool. Gives more reason/gameplay to suit AIs and the people wearing them
## Proof Of Testing
image

https://github.com/user-attachments/assets/015b9e9d-8f30-482c-ba8a-c41f05fc466a
<details>
<summary>Screenshots/Videos</summary>

</details>

## Changelog
:cl:
add: adds a new mod module
/:cl:

### CODE REVIEW

**Comment by Odairu** [comment]
> Can we get a button to disable modsuit AI telemetry while we're at it so people can make MODsuit AIs for this stuff without messing with other things

out of scope, also a pain in the ass to do (since I had to do that while making my antag, and it'd be harder for a non-antag AI)

**Comment by aKromatopzia** [comment]
> > Can we get a button to disable modsuit AI telemetry while we're at it so people can make MODsuit AIs for this stuff without messing with other things
> 
> out of scope, also a pain in the ass to do (since I had to do that while making my antag, and it'd be harder for a non-antag AI)

Yeah, I ken. Goodness knows we're not getting a full AI code refactor from TG either. But! I suppose since in the end any issues with this kind of stuff would be player/IC issue anyways. Hope certain folk actually cooperate with damage mitigation/control!

**Comment by YakumoChen** [comment]
I'd rather this be limited to emag / traitors only, the crew printing full AIs explicitly to ride in their modsuit is powergaming as it is.

**Comment by sippykot** [comment]
imo if the swapped ai/player talk it should still use their original voice instead of the voice of the other person

File: modular_zubbers/code/modules/research/designs/mechfab_designs.dm [hunk 1]
```diff
@@ -108,3 +108,15 @@
 	category = list(
 		RND_CATEGORY_MECHFAB_CYBORG_MODULES + RND_SUBCATEGORY_MECHFAB_CYBORG_MODULES_RESEARCH
 	)
+
+/datum/design/module/mind_transfer
+	name = "Mind Transference Module"
+	id = "mod_mind_transfer"
+	materials = list(
+		/datum/material/iron =HALF_SHEET_MATERIAL_AMOUNT,
+		/datum/material/diamond =SMALL_MATERIAL_AMOUNT*5,
```
Comment (Arturlang) [inline modular_zubbers/code/modules/research/designs/mechfab_designs.dm:117]: While indendation in this file is inconsistent, avoid contributing to it, put a space on each side of the =

**Review by Arturlang — CHANGES_REQUESTED** [review]

**Comment by shellspeed1** [comment]
Actually, I like this idea  but the AI needs to take over the mod specifically with no ability for the player to mess with it imo.

**Comment by Swiftfeather** [comment]
This is bugged in such a way that a player with no AI in the MOD, with the mind transfer module in the MOD, gets ghosted when depowering the suit. I dragged them back into their body but they were also unable to remove the module afterwards, it required a vv delete.

File: modular_zubbers/code/modules/mod/modules.dm [hunk 2]
```diff
@@ -55,3 +55,59 @@
...
+	var/controller = WEARER_CONTROL
+	//Ckey of the original AI
+	var/ai_key
+	//Ckey of the original wearer
+	var/wearer_key
+	cooldown_time = 30 SECONDS
+
+/obj/item/mod/module/mind_swap/on_select()
+	if(!mod.ai_assistant)
+		balloon_alert(mod.wearer, "No AI present for swap.")
```
Comment (Majkl-J) [inline modular_zubbers/code/modules/mod/modules.dm:78]: balloon alerts are not capitalized, and meant to be short
```suggestion
		balloon_alert(mod.wearer, "no AI present")
```

File: modular_zubbers/code/modules/mod/modules.dm [hunk 3]
```diff
@@ -55,3 +55,59 @@
...
+/obj/item/mod/module/mind_swap/on_activation()
+	swap_minds()
+
+/obj/item/mod/module/mind_swap/on_deactivation(display_message, deleting)
+	swap_minds()
+
+/obj/item/mod/module/mind_swap/on_suit_activation()
+
+	ai_key = mod.ai_assistant?.key
+	wearer_key = mod.wearer.key
```
Comment (Majkl-J) [inline modular_zubbers/code/modules/mod/modules.dm:91]:
```suggestion
/obj/item/mod/module/mind_swap/on_suit_activation()
	ai_key = mod.ai_assistant?.key
	wearer_key = mod.wearer.key
```

File: modular_zubbers/code/modules/mod/modules.dm [hunk 4]
```diff
@@ -55,3 +55,59 @@
...
+#define AI_CONTROL 1
+
+/obj/item/mod/module/mind_swap
+	name = "MOD Neural Transference module"
+	desc = "Swaps the MOD wearer's and Assistant AI's neural pathways."
+	removable = FALSE
+	required_slots = list(ITEM_SLOT_FEET, ITEM_SLOT_GLOVES, ITEM_SLOT_OCLOTHING, ITEM_SLOT_HEAD)
+	module_type = MODULE_ACTIVE
+	//Who's in control of the wearer's body
+	var/controller = WEARER_CONTROL
```
Comment (Majkl-J) [inline modular_zubbers/code/modules/mod/modules.dm:69]: you could make this into a boolean, something like ai_control = FALSE, makes the switch logic lower down easier as well.

File: modular_zubbers/code/modules/mod/modules.dm [hunk 5]
```diff
@@ -55,3 +55,59 @@
...
+			mod.ai_assistant.ghostize(FALSE)
+			mod.wearer.key = ai_key
+			mod.ai_assistant.key = wearer_key
+			controller = AI_CONTROL
+		if(AI_CONTROL)
+			mod.wearer.ghostize(FALSE)
+			mod.ai_assistant.ghostize(FALSE)
+			mod.wearer.key = wearer_key
+			mod.ai_assistant.key = ai_key
+			controller = WEARER_CONTROL
```
Comment (Majkl-J) [inline modular_zubbers/code/modules/mod/modules.dm:110]: ditto, with the change suggested above you can simplify this to

```suggestion
/obj/item/mod/module/mind_swap/proc/swap_minds()
	mod.wearer.ghostize(FALSE)
	mod.ai_assistant.ghostize(FALSE)
	if(ai_control)
		mod.wearer.key = wearer_key
		mod.ai_assistant.key = ai_key
	else
		mod.wearer.key = ai_key
		mod.ai_assistant.key = wearer_key
	ai_control = !ai_control
```

**Review by Maia-J — CHANGES_REQUESTED** [review]

**Review by Maia-J — APPROVED** [review]

**Comment by aKromatopzia** [comment]
This has been used to give crew members AI shells, and AI organic bodies. You can argue it's a "drawback" to having no cameras, but wasn't "AI shouldn't have hands" something that's supposed to be held dearly with how much the research borg PR got gutted for it?

**Comment by Odairu** [comment]
> This has been used to give crew members AI shells, and AI organic bodies. You can argue it's a "drawback" to having no cameras, but wasn't "AI shouldn't have hands" something that's supposed to be held dearly with how much the research borg PR got gutted for it?
> 
> Anyways, I must ask for this to be re-reviewed in some way because of that fact.

Its not like the organic can just use the shell whenever, you're not suddenly getting 2 AIs or an AI with hands since, anything the AI can do in the host body, the host can also do themself, and the AI can only do these things so long as they're in the suit with the host, you're not suddenly getting an extra pair of hands, you have two people who can now fill knowledge gaps the other doesn't, say, something engineering related needs done, the carbon doesn't know how, AI can swap and do it, they can't act independently of one another because they're tied by the suit, suit turns off, they switch back etc.

**Comment by aKromatopzia** [comment]
> Its not like the organic can just use the shell whenever, you're not suddenly getting 2 AIs or an AI with hands since, anything the AI can do in the host body, the host can also do themself, and the AI can only do these things so long as they're in the suit with the host, you're not suddenly getting an extra pair of hands, you have two people who can now fill knowledge gaps the other doesn't, say, something engineering related needs done, the carbon doesn't know how, AI can swap and do it, they can't act independently of one another because they're tied by the suit, suit turns off, they switch back etc.

It was one person put into the MODsuit to puppet their own lifeless corpse around.

**Comment by TattoMoon** [comment]
> > This has been used to give crew members AI shells, and AI organic bodies. You can argue it's a "drawback" to having no cameras, but wasn't "AI shouldn't have hands" something that's supposed to be held dearly with how much the research borg PR got gutted for it?
> > Anyways, I must ask for this to be re-reviewed in some way because of that fact.
> 
> Its not like the organic can just use the shell whenever, you're not suddenly getting 2 AIs or an AI with hands since, anything the AI can do in the host body, the host can also do themself, and the AI can only do these things so long as they're in the suit with the host, you're not suddenly getting an extra pair of hands, you have two people who can now fill knowledge gaps the other doesn't, say, something engineering related needs done, the carbon doesn't know how, AI can swap and do it, they can't act independently of one another because they're tied by the suit, suit turns off, they switch back etc.

Yeah so, watching the rounds recently I'm seeing a lot of people getting debrained and using the mind-swap module without permission so.. you have just introduced a way for people to turn themselves into an Ai and become practically unkillable, with most of their reasons being "Lol, mechanic go brrr."
I know from a previous PR, specifically the research cyborgs that one of the issues was "Ai with hands" and now we, do infact, have "AI with hands" as a reality. I think this PR as is, is not great for the server, and I should not have to explain why this isn't great. At least make this locked behind the traitor uplink or something, and not under illegal tech, please. I like the idea but having it be illegal tech is just a terrible idea, and is being proven to be a terrible idea.

**Comment by aKromatopzia** [comment]
Thinking on it a bit more the issue might be at least alleviated by making the swap require a cKey in the body being swapped into

**Pull request merged**